### ISSUE TITLE
Pagination

### ISSUE THREAD

**Issue by aristippe** [aristippe/pathagar#33]
There's various examples of pagination out there, some better than others. I found a simple one and in b47efc147a2cf75e1a596e64c6f2035b27a9dff8, I committed to master since it is simple and doesn't break anything. Feel free to play around with, and someday we can find a better one. What's there is good enough for now, so something better can be done whenever. If someday someone thought too something else would be cool to have, a quick eval of what looks decent might help, and I can see if I can adapt it.

**Comment by sinergatis** [comment]
Pagination is always good, and I did not check other views, although the original `_add_book` seemed to already have pagination in place via a proper `Paginator`? _Unless I'm missing something_, since all your changes on the view seem to be geared towards calculating `page_range` ... can you please revert them and use the [standard Paginator attribute](https://docs.djangoproject.com/en/dev/topics/pagination/#django.core.paginator.Paginator.page_range) in the template? :no_mouth:

**Comment by sinergatis** [comment]
Also, while you are at it it is probably a good idea to remove the `q_count` from the returned context, and use `{{ q.count }}` on the template in its place - it feels cleaner that way (and using `len()` on a `QuerySet` [kind of negates](https://docs.djangoproject.com/en/dev/ref/models/querysets/#count) some of the benefits of the `Paginator` in this case, as it causes _all_ the `Book`s to be retrieved instead of just the ones on the current page).

**Comment by aristippe** [comment]
Unless I'm missing something else, and I did try it a few days ago, unless I missed something there too, page_range will return an interator from 1 to the entire number of pages, correct? As the items per page is currently 50, if there's 10,000 books, the iterator will be from 1 to 200. The pagination I'm referring to is for book_list. For instance Google something and however number of results there are, the paginator at the bottom shows page links for the 5 pages around it:

<img width="432" alt="capture d ecran 2016-02-12 a 08 11 31" src="https://cloud.githubusercontent.com/assets/13608624/13012625/de94c396-d160-11e5-96d0-a475c736978e.png">

That's what the paginator is for. :) Something like this example:

https://www.technovelty.org/web/skipping-pages-with-djangocorepaginator.html

If you were thinking something else looks cleaner, I agree a custom paginator should eventually be a class, though I was inserting something helpful for the moment, as I test around, and was hoping to improve it by trying some of the more robust examples or apps available.

As for "q" and "q_count", q is the query string, and it's originally used for such things as populating the search box with the search string if there's already an active query. I added a use for when there is a search, it'll show:

> Search: "my search string" (q_count results)

just displaying the total number of search results while all other behavior remains the same. Maybe there is a cleaner way, though I'm not aware of another existing variable that shows the number of search results.

**Comment by sinergatis** [comment]
Oh, got it now! Didn't realize you were trying to limit the number of pages shown to the user on the navigation stuff: clearly the  "unless I'm missing something" applies :wine_glass: !

It makes sense, although for reusability I think it is candidate for eventually including it as a [custom template tag](https://docs.djangoproject.com/en/1.9/howto/custom-template-tags/): basically, a tiny function that accepts a `Paginator` and performs that calculations, returning the desired slice of pages. Subclassing `Paginator` would work too, but feels like overkill a bit. It has the plus of getting you to explore custom template tags :trollface: 

As for `q` and `q_count`, sorry, it was a typo on my end - perhaps using `{{ paginator.count }}`, as per the [docs](https://docs.djangoproject.com/en/dev/topics/pagination/#django.core.paginator.Paginator.count), does the trick? I _personally_ like to keep the amount of objects on the context simple, specially in the case where they are easily derived from objects already passed, such as this case (and again, please be careful when using `len()` with a QuerySet - in the scenario above, it would mean fetching all the 10,000 books from the DB every time the view is accessed!).

**Comment by aristippe** [comment]
A custom template tag sounds like a good idea. I've looked briefly over some other's pagination code though I haven't yet tried to see if I can integrate it, into a separate class if needed as most others seem to do. If a custom tag is possible, I'll try.

Sorry for possibly polluting master. :) True it's something that will eventually change; I'm not sure when I'd like to get to it, although having something there I found useful at the moment. If you prefer a branch, I could move it there.

As for q_count, paginator.count I wasn't aware of and it works nicely. A change for that is coming shortly.
